### PR TITLE
ebuild-writing/misc-files/patches: Add section about clean patches

### PR DESCRIPTION
**Pull request by ulm**
Taken from https://dev.gentoo.org/~vapier/clean-patches, with some rewording.
(Work in progress.)

Bug: https://bugs.gentoo.org/247266

### CODE REVIEW

File: ebuild-writing/misc-files/patches/text.xml [hunk 1]
```diff
@@ -109,5 +109,277 @@ from the <c>vim</c> patch tarball:
...
+      <li>Who fixed the bug?</li>
+      <li>Who wrote the patch?</li>
+      <li>Who tested the patch?</li>
+      <li>Who gave advice on the patch?</li>
+    </ul>
+  </li>
+</ul>
+
+<p>
+All this information should be <e>in the patch itself</e>. It should never ever
```
Comment (thesamesam) [inline ebuild-writing/misc-files/patches/text.xml:224]:
```suggestion
All this information should be <e>in the patch itself</e>. It should never
```

File: ebuild-writing/misc-files/patches/text.xml [hunk 2]
```diff
@@ -109,5 +109,277 @@ from the <c>vim</c> patch tarball:
...
+be found in something like the ebuild. If you really really want to put a
```
Comment (thesamesam) [inline ebuild-writing/misc-files/patches/text.xml:225]:
```suggestion
be found in something like the ebuild. If you really want to put a
```

or

```suggestion
be found in something like the ebuild. If you must put a
```

File: ebuild-writing/misc-files/patches/text.xml (hunk 1 above)
Comment (thesamesam) [inline ebuild-writing/misc-files/patches/text.xml:224]:
```suggestion
All this information should be (at least) <e>in the patch itself</e>. It should never ever
```

File: ebuild-writing/misc-files/patches/text.xml [hunk 3]
```diff
@@ -109,5 +109,277 @@ from the <c>vim</c> patch tarball:
...
+<pre>
+From: Nice Person &lt;foo@cow.example.com&gt;
+</pre>
+
+<p>
+Or when documenting relevant URLs, use something like:
+</p>
+
+<pre>
+Bug: https://upstream.tracker.com/12345
```
Comment (thesamesam) [inline ebuild-writing/misc-files/patches/text.xml:248]:
```suggestion
Bug: https://upstream.example.com/12345
```

File: ebuild-writing/misc-files/patches/text.xml [hunk 4]
```diff
@@ -109,5 +109,277 @@ from the <c>vim</c> patch tarball:
...
+</p>
+
+<p>
+Also note that <c>patch</c> uses the timestamp info in order to remove empty
+files automatically. Alternatively, you can specify the <c>-E</c> option with
+<c>eapply</c> if you want to remove an empty file.
+</p>
+
+<p>
+The following script will help deleting these things:
```
Comment (thesamesam) [inline ebuild-writing/misc-files/patches/text.xml:292]: Maybe note that people shouldn't do this in the ebuild, although it's obvious.

File: ebuild-writing/misc-files/patches/text.xml [hunk 5]
```diff
@@ -109,5 +109,277 @@ from the <c>vim</c> patch tarball:
...
+
+<p>
+All this information should be <e>in the patch itself</e>. It should never ever
+be found in something like the ebuild. If you really really want to put a
+comment next to a patch in an ebuild, then this is about the only thing that is
+OK (where 12345 is the Gentoo bug number):
+</p>
+
+<codesample lang="ebuild">
+eapply "${FILESDIR}"/${P}-fatty-cow.patch #12345
```
[the remark this answers is not in the document]
Comment (SoapGentoo) [reply ebuild-writing/misc-files/patches/text.xml:231]: agreed

File: ebuild-writing/misc-files/patches/text.xml [hunk 6]
```diff
@@ -109,5 +109,277 @@ from the <c>vim</c> patch tarball:
...
+                  ^^^^^^^^^^^^ keep this part
+</pre>
+
+<p>
+Extra points if you make the filename in the <c>---</c>/<c>+++</c> section
+consistent and sane. That is, remove different leading <c>backup/paths/</c>
+and <c>.orig</c>/<c>.new</c> suffixes. Your patch should be in the <c>-p1</c>
+format because this tends to be much more standard than any other <c>-p#</c>.
+It is also what <c>eapply</c> understands by default. A good suggestion is to
+use the package name/version as the leading portion that gets stripped.
```
[the remark this answers is not in the document]
Comment (SoapGentoo) [reply ebuild-writing/misc-files/patches/text.xml:282]: `a/` and `b/` are more forward-agnostic, so I would go with this too.

**Comment by SoapGentoo** [comment]
I agree with all of @thesamesam's suggestions, let's get those in.

File: ebuild-writing/misc-files/patches/text.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml]: I've mentioned both alternatives.

File: ebuild-writing/misc-files/patches/text.xml (hunk 1 above)
Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml]: I haven't included this one.

File: ebuild-writing/misc-files/patches/text.xml (hunk 4 above)
Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml:295]: Added a `#!/bin/bash` shebang which should make it clear.

File: ebuild-writing/misc-files/patches/text.xml (hunk 1 above)
Comment (thesamesam) [reply ebuild-writing/misc-files/patches/text.xml]: No problem.

File: ebuild-writing/misc-files/patches/text.xml (hunk 4 above)
Comment (thesamesam) [reply ebuild-writing/misc-files/patches/text.xml:295]: Sure.

File: ebuild-writing/misc-files/patches/text.xml [hunk 7]
```diff
@@ -109,5 +109,281 @@ from the <c>vim</c> patch tarball:
...
+<p>
+All this information should be <e>in the patch itself</e>. It should never be
+found in something like the ebuild. If you really want to put a comment next
+to a patch in an ebuild, then this is about the only thing that is OK
+(where 12345 is the Gentoo bug number):
+</p>
+
+<codesample lang="ebuild">
+PATCHES=(
+	"${FILESDIR}"/${P}-fatty-cow.patch #12345
```
Comment (thesamesam) [inline ebuild-writing/misc-files/patches/text.xml:232]: Let's do one more: different patch name, given we changed the others?

**Review by thesamesam — APPROVED** [review]

File: ebuild-writing/misc-files/patches/text.xml [hunk 8]
```diff
@@ -109,5 +109,282 @@ from the <c>vim</c> patch tarball:
...
+<p>
+And if you want to note your copyright signoff, slap on a Signed-off-by tag:
+</p>
+
+<pre>
+Signed-off-by: Diligent Developer &lt;larry@gentoo.org&gt;
+</pre>
+
+<p>
+Finally, your patch should be clear of useless cruft, including dates. If it
```
Comment (gktrk) [inline ebuild-writing/misc-files/patches/text.xml:263]: Is the `dates` different from `the timestamps on the files` 4 lines below?

**Review by gktrk — COMMENTED** [review]
Other than the one comment I have, I'm good with all the changes.

File: ebuild-writing/misc-files/patches/text.xml (hunk 8 above)
Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml]: The "dates" were added in a previous round of review. Changed to "timestamps" now.

Comment (gktrk) [reply ebuild-writing/misc-files/patches/text.xml]: My point rather was that if we mention it 4 lines below in a more meaningful context, is there a value to mentioning it here? `useless cruft` is ambiguous anyway, which is why it needs the sentences that follow it for clarification. Simply adding `, including timestamps`, only to mention it again in more detail later, doesn't do a good job helping clarify the ambiguity IMO.

Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml]: Right, I misunderstood that. Removed now.

File: ebuild-writing/misc-files/patches/text.xml [hunk 9]
```diff
@@ -109,5 +109,282 @@ from the <c>vim</c> patch tarball:
...
+
+<ul>
+  <li>
+    External references
+    <ul>
+      <li>Upstream mailing archives</li>
+      <li>Upstream bug reports</li>
+      <li>Upstream commit links</li>
+      <li>upstream changelog entries</li>
+      <li>Gentoo bug reports</li>
```
Comment (juippis) [inline ebuild-writing/misc-files/patches/text.xml:179]: One lowercase `u` there.

File: ebuild-writing/misc-files/patches/text.xml [hunk 10]
```diff
@@ -109,5 +109,282 @@ from the <c>vim</c> patch tarball:
...
+
+<subsection>
+<title>Why</title>
+<body>
+
+<p>
+This may take more effort "up front", but the amount of effort that it saves
+for everyone else in the future more than makes up for it. This refers to other
+distributions or upstream maintainers who read the patch, or future Gentoo
+maintainers/developers. By keeping all patches "clean", people can quickly and
```
Comment (juippis) [inline ebuild-writing/misc-files/patches/text.xml:133]: `Gentoo maintainers/developers` does this need to be separated? Aren't package maintainers called maintainers regardless of their status?

File: ebuild-writing/misc-files/patches/text.xml [hunk 11]
```diff
@@ -109,5 +109,282 @@ from the <c>vim</c> patch tarball:
...
+<pre>
+@@ -80,6 +82,7 @@ case $sys in
+                  ^^^^^^^^^^^^ keep this part
+</pre>
+
+<p>
+Extra points if you make the filename in the <c>---</c>/<c>+++</c> section
+consistent and sane. That is, remove different leading <c>backup/paths/</c>
+and <c>.orig</c>/<c>.new</c> suffixes. Your patch should be in the <c>-p1</c>
+format because this tends to be much more standard than any other <c>-p#</c>.
```
Comment (juippis) [inline ebuild-writing/misc-files/patches/text.xml:282]: I'd say to be more confident in here and say it is more standard than the others, especially to unify major Linux distributions in patch sharing.

File: ebuild-writing/misc-files/patches/text.xml (hunk 10 above)
Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml]: Yep, "maintainers" is enough here.

File: ebuild-writing/misc-files/patches/text.xml (hunk 11 above)
Comment (ulm) [reply ebuild-writing/misc-files/patches/text.xml:282]: I think the wording is fine as-is. We mention the eapply default in the next sentence, so there should be enough rationale.

**Pull request merged**